src/snowflake/connector/connection.py

```diff
@@ -1981,3 +1981,23 @@ def _log_telemetry_imported_packages(self) -> None:
                     connection=self,
                 )
             )
+
+    def is_valid(self) -> bool:
+        """This function tries to answer the question: Is this connection still good for sending queries?
+        Attempts to validate the connections both on the TCP/IP and Session levels."""
+        logger.debug("validating connection and session")
+        if self.is_closed():
+            logger.debug("connection is already closed and not valid")
+            return False
+        else:
+            try:
+                logger.debug("trying to heartbeat into the session to validate")
+                hb_result = self.rest._heartbeat()
+                session_valid = hb_result.get("success")
+                logger.debug("session still valid? %s", str(session_valid))
+                return session_valid
+            except Exception as e:
+                logger.debug(
+                    "session could not be validated due to exception: %s", str(e)
```

Comment:
Th `%s` should do this automatically
```suggestion
                    "session could not be validated due to exception: %s", e
```